File: JsonSchemaAndStructs/KiteConnect-Python-DataClass/basket_margins.py
# To use this code, make sure you
#
#     import json
#
# and then, to convert JSON from a string, do
#
#     result = basket_margins_from_dict(json.loads(json_string))

from dataclasses import dataclass
from typing import Optional, Any, List, TypeVar, Type, cast, Callable


T = TypeVar("T")


def from_int(x: Any) -> int:
    assert isinstance(x, int) and not isinstance(x, bool)
    return x


def from_none(x: Any) -> Any:
    assert x is None
    return x


def from_union(fs, x):
    for f in fs:
        try:
            return f(x)
        except:
            pass
    assert False


def from_str(x: Any) -> str:
    assert isinstance(x, str)
    return x


def from_float(x: Any) -> float:
    assert isinstance(x, (float, int)) and not isinstance(x, bool)
    return float(x)


def to_float(x: Any) -> float:
    assert isinstance(x, float)
    return x


def to_class(c: Type[T], x: Any) -> dict:
    assert isinstance(x, c)
    return cast(Any, x).to_dict()


def from_list(f: Callable[[Any], T], x: Any) -> List[T]:
    assert isinstance(x, list)
    return [f(y) for y in x]


@dataclass(slots=True)
class Pnl:
    realised: Optional[int] = None
    unrealised: Optional[int] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Pnl':
        assert isinstance(obj, dict)
        realised = from_union([from_int, from_none], obj.get("realised"))
        unrealised = from_union([from_int, from_none], obj.get("unrealised"))
        return Pnl(realised, unrealised)

    def to_dict(self) -> dict:
        result: dict = {}
        result["realised"] = from_union([from_int, from_none], self.realised)
        result["unrealised"] = from_union([from_int, from_none], self.unrealised)
        return result


@dataclass(slots=True)
class Final:
    additional: Optional[int] = None
    bo: Optional[int] = None
    cash: Optional[int] = None
    exchange: Optional[str] = None
    exposure: Optional[float] = None
    option_premium: Optional[float] = None
    pnl: Optional[Pnl] = None
    span: Optional[float] = None
    total: Optional[float] = None
    tradingsymbol: Optional[str] = None
    type: Optional[str] = None
    var: Optional[int] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Final':
        assert isinstance(obj, dict)
        additional = from_union([from_int, from_none], obj.get("additional"))
        bo = from_union([from_int, from_none], obj.get("bo"))
        cash = from_union([from_int, from_none], obj.get("cash"))
        exchange = from_union([from_str, from_none], obj.get("exchange"))
        exposure = from_union([from_float, from_none], obj.get("exposure"))
        option_premium = from_union([from_float, from_none], obj.get("option_premium"))
        pnl = from_union([Pnl.from_dict, from_none], obj.get("pnl"))
        span = from_union([from_float, from_none], obj.get("span"))
        total = from_union([from_float, from_none], obj.get("total"))
        tradingsymbol = from_union([from_str, from_none], obj.get("tradingsymbol"))
        type = from_union([from_str, from_none], obj.get("type"))
        var = from_union([from_int, from_none], obj.get("var"))
        return Final(additional, bo, cash, exchange, exposure, option_premium, pnl, span, total, tradingsymbol, type, var)

    def to_dict(self) -> dict:
        result: dict = {}
        result["additional"] = from_union([from_int, from_none], self.additional)
        result["bo"] = from_union([from_int, from_none], self.bo)
        result["cash"] = from_union([from_int, from_none], self.cash)
        result["exchange"] = from_union([from_str, from_none], self.exchange)
        result["exposure"] = from_union([to_float, from_none], self.exposure)
        result["option_premium"] = from_union([to_float, from_none], self.option_premium)
        result["pnl"] = from_union([lambda x: to_class(Pnl, x), from_none], self.pnl)
        result["span"] = from_union([to_float, from_none], self.span)
        result["total"] = from_union([to_float, from_none], self.total)
        result["tradingsymbol"] = from_union([from_str, from_none], self.tradingsymbol)
        result["type"] = from_union([from_str, from_none], self.type)
        result["var"] = from_union([from_int, from_none], self.var)
        return result


@dataclass(slots=True)
class Data:
    final: Optional[Final] = None
    initial: Optional[Final] = None
    orders: Optional[List[Final]] = None

    @staticmethod
    def from_dict(obj: Any) -> 'Data':
        assert isinstance(obj, dict)
        final = from_union([Final.from_dict, from_none], obj.get("final"))
        initial = from_union([Final.from_dict, from_none], obj.get("initial"))
        orders = from_union([lambda x: from_list(Final.from_dict, x), from_none], obj.get("orders"))
        return Data(final, initial, orders)

    def to_dict(self) -> dict:
        result: dict = {}
        result["final"] = from_union([lambda x: to_class(Final, x), from_none], self.final)
        result["initial"] = from_union([lambda x: to_class(Final, x), from_none], self.initial)
        result["orders"] = from_union([lambda x: from_list(lambda x: to_class(Final, x), x), from_none], self.orders)
        return result


@dataclass(slots=True)
class BasketMargins:
    data: Optional[Data] = None
    status: Optional[str] = None

    @staticmethod
    def from_dict(obj: Any) -> 'BasketMargins':
        assert isinstance(obj, dict)
        data = from_union([Data.from_dict, from_none], obj.get("data"))
        status = from_union([from_str, from_none], obj.get("status"))
        return BasketMargins(data, status)

    def to_dict(self) -> dict:
        result: dict = {}
        result["data"] = from_union([lambda x: to_class(Data, x), from_none], self.data)
        result["status"] = from_union([from_str, from_none], self.status)
        return result


def basket_margins_from_dict(s: Any) -> BasketMargins:
    return BasketMargins.from_dict(s)


def basket_margins_to_dict(x: BasketMargins) -> Any:
    return to_class(BasketMargins, x)

File: JsonSchemaAndStructs/KiteConnect-JavaScript-Structs/basket_margins.js
// To parse this data:
//
//   const Convert = require("./file");
//
//   const basketMargins = Convert.toBasketMargins(json);
//
// These functions will throw an error if the JSON doesn't
// match the expected interface, even if the JSON is valid.

// Converts JSON strings to/from your types
// and asserts the results of JSON.parse at runtime
function toBasketMargins(json) {
    return cast(JSON.parse(json), r("BasketMargins"));
}

function basketMarginsToJson(value) {
    return JSON.stringify(uncast(value, r("BasketMargins")), null, 2);
}

function invalidValue(typ, val, key = '') {
    if (key) {
        throw Error(`Invalid value for key "${key}". Expected type ${JSON.stringify(typ)} but got ${JSON.stringify(val)}`);
    }
    throw Error(`Invalid value ${JSON.stringify(val)} for type ${JSON.stringify(typ)}`, );
}

function jsonToJSProps(typ) {
    if (typ.jsonToJS === undefined) {
        const map = {};
        typ.props.forEach((p) => map[p.json] = { key: p.js, typ: p.typ });
        typ.jsonToJS = map;
    }
    return typ.jsonToJS;
}

function jsToJSONProps(typ) {
    if (typ.jsToJSON === undefined) {
        const map = {};
        typ.props.forEach((p) => map[p.js] = { key: p.json, typ: p.typ });
        typ.jsToJSON = map;
    }
    return typ.jsToJSON;
}

function transform(val, typ, getProps, key = '') {
    function transformPrimitive(typ, val) {
        if (typeof typ === typeof val) return val;
        return invalidValue(typ, val, key);
    }

    function transformUnion(typs, val) {
        // val must validate against one typ in typs
        const l = typs.length;
        for (let i = 0; i < l; i++) {
            const typ = typs[i];
            try {
                return transform(val, typ, getProps);
            } catch (_) {}
        }
        return invalidValue(typs, val);
    }

    function transformEnum(cases, val) {
        if (cases.indexOf(val) !== -1) return val;
        return invalidValue(cases, val);
    }

    function transformArray(typ, val) {
        // val must be an array with no invalid elements
        if (!Array.isArray(val)) return invalidValue("array", val);
        return val.map(el => transform(el, typ, getProps));
    }

    function transformDate(val) {
        if (val === null) {
            return null;
        }
        const d = new Date(val);
        if (isNaN(d.valueOf())) {
            return invalidValue("Date", val);
        }
        return d;
    }

    function transformObject(props, additional, val) {
        if (val === null || typeof val !== "object" || Array.isArray(val)) {
            return invalidValue("object", val);
        }
        const result = {};
        Object.getOwnPropertyNames(props).forEach(key => {
            const prop = props[key];
            const v = Object.prototype.hasOwnProperty.call(val, key) ? val[key] : undefined;
            result[prop.key] = transform(v, prop.typ, getProps, prop.key);
        });
        Object.getOwnPropertyNames(val).forEach(key => {
            if (!Object.prototype.hasOwnProperty.call(props, key)) {
                result[key] = transform(val[key], additional, getProps, key);
            }
        });
        return result;
    }

    if (typ === "any") return val;
    if (typ === null) {
        if (val === null) return val;
        return invalidValue(typ, val);
    }
    if (typ === false) return invalidValue(typ, val);
    while (typeof typ === "object" && typ.ref !== undefined) {
        typ = typeMap[typ.ref];
    }
    if (Array.isArray(typ)) return transformEnum(typ, val);
    if (typeof typ === "object") {
        return typ.hasOwnProperty("unionMembers") ? transformUnion(typ.unionMembers, val)
            : typ.hasOwnProperty("arrayItems")    ? transformArray(typ.arrayItems, val)
            : typ.hasOwnProperty("props")         ? transformObject(getProps(typ), typ.additional, val)
            : invalidValue(typ, val);
    }
    // Numbers can be parsed by Date but shouldn't be.
    if (typ === Date && typeof val !== "number") return transformDate(val);
    return transformPrimitive(typ, val);
}

function cast(val, typ) {
    return transform(val, typ, jsonToJSProps);
}

function uncast(val, typ) {
    return transform(val, typ, jsToJSONProps);
}

function a(typ) {
    return { arrayItems: typ };
}

function u(...typs) {
    return { unionMembers: typs };
}

function o(props, additional) {
    return { props, additional };
}

function m(additional) {
    return { props: [], additional };
}

function r(name) {
    return { ref: name };
}

const typeMap = {
    "BasketMargins": o([
        { json: "data", js: "data", typ: u(undefined, r("Data")) },
        { json: "status", js: "status", typ: u(undefined, "") },
    ], false),
    "Data": o([
        { json: "final", js: "final", typ: u(undefined, r("Final")) },
        { json: "initial", js: "initial", typ: u(undefined, r("Final")) },
        { json: "orders", js: "orders", typ: u(undefined, a(r("Final"))) },
    ], false),
    "Final": o([
        { json: "additional", js: "additional", typ: u(undefined, 0) },
        { json: "bo", js: "bo", typ: u(undefined, 0) },
        { json: "cash", js: "cash", typ: u(undefined, 0) },
        { json: "exchange", js: "exchange", typ: u(undefined, "") },
        { json: "exposure", js: "exposure", typ: u(undefined, 3.14) },
        { json: "option_premium", js: "option_premium", typ: u(undefined, 3.14) },
        { json: "pnl", js: "pnl", typ: u(undefined, r("Pnl")) },
        { json: "span", js: "span", typ: u(undefined, 3.14) },
        { json: "total", js: "total", typ: u(undefined, 3.14) },
        { json: "tradingsymbol", js: "tradingsymbol", typ: u(undefined, "") },
        { json: "type", js: "type", typ: u(undefined, "") },
        { json: "var", js: "var", typ: u(undefined, 0) },
    ], false),
    "Pnl": o([
        { json: "realised", js: "realised", typ: u(undefined, 0) },
        { json: "unrealised", js: "unrealised", typ: u(undefined, 0) },
    ], false),
};

module.exports = {
    "basketMarginsToJson": basketMarginsToJson,
    "toBasketMargins": toBasketMargins,
};
